### PR TITLE
Unit test for FactOfTheDayService

### PR DESCRIPTION
**Pull request by SashkoMolodec**

### CODE REVIEW

File: core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+import static org.mockito.Mockito.*;
+import org.mockito.MockitoAnnotations;
+import org.modelmapper.ModelMapper;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+@ExtendWith(SpringExtension.class)
```
Comment (antonbakalets) [inline core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java:33]: You don't need SpringExtension as spring context is not used in the test, MockitoExtension would be enough.

File: core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+public class FactOfTheDayServiceImplTest {
+    @InjectMocks
+    private FactOfTheDayServiceImpl factOfTheDayService;
```
Comment (antonbakalets) [inline core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java:36]: The field annotated with `@InjectMocks` should be the last one. When I read the code, I want to see all the `@Mock`-ed objects and then at the end the object that those mocks are injected in.

File: core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+    @Mock
+    private FactOfTheDayRepo factOfTheDayRepo;
+
+    @Mock
+    private LanguageService languageService;
+
+    @BeforeEach
+    public void setup(){
+        MockitoAnnotations.initMocks(this);
```
Comment (antonbakalets) [inline core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java:52]: You don't need this as both `SpringExtension` and `MockitoExtension `are doing this for you under the hood.

**Review by antonbakalets — CHANGES_REQUESTED** [review]

File: core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java (hunk 3 above)
Comment (SashkoMolodec) [reply core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java:52]: Знімок екрана 2020-09-11 084226

Comment (SashkoMolodec) [reply core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java:52]: For some reason it doesn't

Comment (SashkoMolodec) [reply core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java:52]: Objects that aren't calling directly in test class are not getting mocked

File: core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java (hunk 1 above)
Comment (SashkoMolodec) [reply core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java]: Changes done.

File: core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java (hunk 2 above)
Comment (SashkoMolodec) [reply core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java]: Changes done.

File: core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java (hunk 3 above)
Comment (antonbakalets) [reply core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java:53]: I've found the problem. Part of fields in `FactOfTheDayServiceImpl` are initialized via constructor, another part via `@Autovired` annotation. `Mockito` cannot handle both. Need a consistent approach. I suggest, `FactOfTheDayServiceImpl` constructor should be removed.

Comment (SashkoMolodec) [reply core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java:53]: So should I close this request change? Or make changes in FactOfTheDayServiceImpl?

Comment (antonbakalets) [reply core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java:53]: Writing unit tests sometime requires target class refactoring. Especially, when the unit test discovers a problem in the target class. Please update `FactOfTheDayServiceImpl`.

Comment (SashkoMolodec) [reply core/src/test/java/greencity/service/impl/FactOfTheDayServiceImplTest.java]: Changes done.

**Comment by SashkoMolodec** [comment]
image

**Pull request closed without merging**